index.js

```diff
@@ -1,87 +1,115 @@
-      AND "memory" = '${query.memory}'
-      AND "test" = '${query.test}'
-      AND ${query.timeFilter}
-      GROUP BY * fill(none)`;
-  },
-  mtbf: (query) => {
-    return `SELECT SUM(value) / SUM(failures) as value
-      FROM "mtbf"
-      WHERE "node" = '${query.node}'
-      AND "branch" = '${query.branch}'
-      AND "device" = '${query.device}'
-      AND "memory" = '${query.memory}'
-      AND ${query.timeFilter}
-      GROUP BY time(1s) fill(none)`;
-  },
-  power: (query) => {
-    return `SELECT value
-      FROM "power"
-      WHERE "test" = '${query.test}'
-      AND "context" = '${query.context}'
-      AND "branch" = '${query.branch}'
-      AND "device" = '${query.device}'
-      AND "memory" = '${query.memory}'
-      AND ${query.timeFilter}`;
-  }
-};
+let R = require('ramda');
+let query = require('./lib');
 
 /**
- * Remove extra whitespace and newlines from a query
- * @param {String} queryString
- * @returns {String}
+ * Determine if an object is a `measure` measurement
+ * @param {Object} options
+ * @returns {Boolean}
+ */
+let isMeasure = R.pipe(R.prop('measurement'), R.equals('measure'));
+
+/**
+ * Determine if an object is a `memory` measurement
+ * @param {Object} options
+ * @returns {Boolean}
  */
-var clean = R.pipe(R.replace(/\n[ ]+/g, ' '), R.trim);
+let isMemory = R.pipe(R.prop('measurement'), R.equals('memory'));
 
 /**
- * Fetch the query string for a given measurement using options to fill in query values
- * @param {String} measurement InfluxDB measurement
- * @param {Object} options query values
+ * Determine if an object is a `measure` or `memory` measurement
+ * @param {Object} options
+ * @returns {Boolean}
  */
-var getRawQuery = (measurement, options) => R.call(R.prop(measurement, queries), options);
+let isMeasureOrMemory = R.converge(R.or, [isMeasure, isMemory]);
 
 /**
- * Fetch the query string for a given set of options
+ * Determine if an object's `context` property is empty
+ * @param {Object} options
+ * @returns {Boolean}
+ */
+let noContext = R.pipe(R.prop('context'), R.converge(R.or, [R.isNil, R.isEmpty]));
+
+/**
+ * Determine if an object's `metric` property is empty
+ * @param {Object} options
+ * @returns {Boolean}
+ */
+let noMetric = R.pipe(R.prop('metric'), R.converge(R.or, [R.isNil, R.isEmpty]));
+
+/**
+ * Determine if an object is missing a required context
+ * @param {Object} options
+ * @returns {Boolean}
+ */
+let missingContext = R.converge(R.and, [isMeasureOrMemory, noContext]);
+
+/**
+ * Determine if an object is missing a required metric
+ * @param {Object} options
+ * @returns {Boolean}
+ */
+let missingMetric = R.converge(R.and, [isMeasureOrMemory, noMetric]);
+
+/**
+ * Return the time filter of an object with a time property
  * @param {Object} options
  * @returns {String}
  */
-var getQuery = R.converge(R.pipe(getRawQuery, clean), [R.prop('measurement'), R.identity]);
+let getTimeFilter = R.pipe(
+  R.prop('time'),
+  R.cond([
+    [R.contains(' '), R.identity],
```

Comment:
Ok cool. Right so the second condition always happens if there's no space in the expression. Just wanted to make sure there would never be a valid expression provided that doesn't have a space, otherwise it would be replaced/expanded out.